csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs

```diff
@@ -0,0 +1,502 @@
+
+        }
+
+        /// <summary>
+        /// Saves a checkpoint to path. It can be loaded into <see cref="CheckpointState"/>
+        /// </summary>
+        /// <param name="path">Specify path for saving the checkpoint.</param>
+        /// <param name="saveOptimizerState">SFlag indicating whether to save optimizer state or not.</param>
+        public void SaveCheckpoint(string path, bool saveOptimizerState = false)
+        {
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtSaveCheckpoint(NativeMethods.GetPlatformSerializedString(path),_nativeHandle, saveOptimizerState));
+        }
+
+        #endregion
+        #region private methods
+
+        private void Init(SessionOptions sessOptions, CheckpointState state, byte[] trainModelPath, byte[] evalModelPath, byte[] optimizerModelPath)
+        {
+            if (!NativeTrainingMethods.TrainingEnabled())
+            {
+                throw new InvalidOperationException("Training is disabled in the current build.");
+            }
+            var options = sessOptions;
+            if (sessOptions == null)
+            {
+                _builtInSessionOptions = new SessionOptions();
+                options = _builtInSessionOptions;
+            }
+            var envHandle = OrtEnv.Handle;
+            try
+            {
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtCreateTrainingSession(envHandle, options.Handle, state.Handle, trainModelPath,
+                                                                                     evalModelPath, optimizerModelPath, out _nativeHandle));
+
+                UIntPtr outputCount = UIntPtr.Zero;
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetTrainModelOutputCount(_nativeHandle, out outputCount));
+                _trainOutputCount = outputCount.ToUInt64();
+
+                // get all the output names and metadata
+                _trainOutputNames = new List<string>();
+                for (ulong i = 0; i < _trainOutputCount; i++)
+                {
+                    _trainOutputNames.Add(GetOutputName(i, true));
+                }
+
+                if (evalModelPath != null)
+                {
+                    outputCount = UIntPtr.Zero;
+                    NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetEvalModelOutputCount(_nativeHandle, out outputCount));
+                    _evalOutputCount = outputCount.ToUInt64();
+                    _evalOutputNames = new List<string>();
+                    for (ulong i = 0; i < _evalOutputCount; i++)
+                    {
+                        _evalOutputNames.Add(GetOutputName(i, false));
+                    }
+                }
+
+                _builtInRunOptions = new RunOptions();  // create a default built-in run option, and avoid creating a new one every run() call
+            } catch
+            {
+                Dispose();
```

Comment:
>  Dispose(); [](http://example.com/codeflow?start=14&length=12)

You should not call Dispose() on the object that is not fully constructed. You can have a helper method that is used both here and in the Dispose().